api/src/main/java/org/eclipse/microprofile/concurrent/ManagedExecutorBuilder.java

```diff
@@ -0,0 +1,115 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.eclipse.microprofile.concurrent;
+
+import org.eclipse.microprofile.concurrent.spi.ConcurrencyProvider;
+
+/**
+ * <p>Builder for <code>ManagedExecutor</code> instances.</p>
+ *
+ * <p>Example usage:</p>
+ * <pre><code> ManagedExecutor executor = ManagedExecutorBuilder.instance()
+ *                                                  .maxAsync(5)
+ *                                                  .maxQueued(20)
+ *                                                  .propagated(ThreadContext.SECURITY)
+ *                                                  .build();
+ * ...
+ * </code></pre>
+ */
+public interface ManagedExecutorBuilder {
+    /**
+     * <p>Builds a new <code>ManagedExecutor</code> with the configuration
+     * that this builder represents as of the point in time when this method
+     * is invoked.</p>
+     *
+     * <p>After <code>build</code> is invoked, the builder instance retains its
+     * configuration and may be further updated to represent different
+     * configurations and build additional <code>ManagedExecutor</code>
+     * instances.</p>
+     *
+     * @return new instance of <code>ManagedExecutor</code>.
+     */
+    ManagedExecutor build();
+
+    /**
+     * <p>Defines the set of thread context types to capture from the thread
+     * that creates a dependent stage (or that submits a task) and which to
+     * propagate to the thread where the action or task executes.</p>
+     *
+     * <p>This set replaces the set that was previously specified on the
+     * builder instance.</p>
+     *
+     * <p>The default set of thread context types is those required by the
+     * EE Concurrency spec, plus CDI.</p>
+     *
+     * <p>Constants for specifying some of the core context types are provided
+     * on {@link ThreadContext}. Other thread context types must be defined
+     * by the specification that defines the context type or by a related
+     * MicroProfile specification.</p>
+     *
+     * <p>Inclusion of a thread context type with prerequisites implies
+     * inclusion of the prerequisites, even if not explicitly specified.</p>
+     *
+     * <p>Thread context types which are not otherwise included in this set
+     * are cleared from the thread of execution for the duration of the
+     * action or task.</p>
+     *
+     * @param types types of thread context to capture and propagate.
+     * @return the same builder instance upon which this method is invoked.
+     */
+    ManagedExecutorBuilder propagated(String... types);
```

Comment:
This is type safe though. Is it possible to pass in ThreadContext type instead and then in ThreadContext to create instances for the contexts

Reply:
It should first be pointed out the builder pattern here exactly mirrors the CDI pattern for defining annotatively.  See the `propagated` method on https://github.com/eclipse/microprofile-concurrency/blob/master/api/src/main/java/org/eclipse/microprofile/concurrent/ManagedExecutorConfig.java, which is used as
`@Inject @ManagedExecutorConfig(propagated = { ThreadContext.SECURITY,  ThreadContext.APPLICATION)`
If this is going to be changed, it should be done on all of ManagedExecutorBuilder, ThreadContextBuilder, ManagedExecutorConfig, and ThreadContextConfig for consistency.
It would be wrong to supply a `ThreadContext` instance to the `ManagedExecutorBuilder.propagated` method because `ThreadContext` also allows for the specification of `unchanged` context which would be non-deterministic and behaviorally incorrect for CompletableFutures, undermining the purpose of this spec.
I'm open to some sort of enumerated type here rather than constants, but we would need to figure out how to do that in a way that is extensible (We want third party thread context providers to define their own context types that can be supplied here).
In any case, I would consider this design point a separate issue given that the changes under this pull are only copying what is already done by the existing `ManagedExecutorConfig`.

Reply:
Issue is: https://github.com/eclipse/microprofile-concurrency/issues/19